worldedit-fabric/src/main/java/com/sk89q/worldedit/fabric/FabricWorld.java

```diff
@@ -0,0 +1,599 @@
+            return info.getRainTime();
+        }
+        return info.getClearWeatherTime();
+    }
+
+    @Override
+    public void setWeather(WeatherType weatherType) {
+        setWeather(weatherType, 0);
+    }
+
+    @Override
+    public void setWeather(WeatherType weatherType, long duration) {
+        LevelProperties info = getWorld().getLevelProperties();
+        if (weatherType == WeatherTypes.THUNDER_STORM) {
+            info.setClearWeatherTime(0);
+            info.setThundering(true);
+            info.setThunderTime((int) duration);
+        } else if (weatherType == WeatherTypes.RAIN) {
+            info.setClearWeatherTime(0);
+            info.setRaining(true);
+            info.setRainTime((int) duration);
+        } else if (weatherType == WeatherTypes.CLEAR) {
+            info.setRaining(false);
+            info.setThundering(false);
+            info.setClearWeatherTime((int) duration);
+        }
+    }
+
+    @Override
+    public int getMaxY() {
+        return getWorld().getHeight() - 1;
+    }
+
+    @Override
+    public BlockVector3 getSpawnPosition() {
+        return FabricAdapter.adapt(getWorld().getSpawnPos());
+    }
+
+    @Override
+    public BlockState getBlock(BlockVector3 position) {
+        net.minecraft.block.BlockState mcState = getWorld()
+                .getChunk(position.getBlockX() >> 4, position.getBlockZ() >> 4)
+                .getBlockState(FabricAdapter.toBlockPos(position));
+
+        BlockState matchingBlock = BlockStateIdAccess.getBlockStateById(Block.getRawIdFromState(mcState));
+        if (matchingBlock != null) {
+            return matchingBlock;
+        }
+
+        return FabricAdapter.adapt(mcState);
+    }
+
+    @Override
+    public BaseBlock getFullBlock(BlockVector3 position) {
+        BlockPos pos = new BlockPos(position.getBlockX(), position.getBlockY(), position.getBlockZ());
+        // Avoid creation by using the CHECK mode -- if it's needed, it'll be re-created anyways
+        BlockEntity tile = ((WorldChunk) getWorld().getChunk(pos)).getBlockEntity(pos, WorldChunk.CreationType.CHECK);
+
+        if (tile != null) {
+            return getBlock(position).toBaseBlock(NBTConverter.fromNative(TileEntityUtils.copyNbtData(tile)));
+        } else {
+            return getBlock(position).toBaseBlock();
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return getWorld().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == null) {
+            return false;
+        } else if ((o instanceof FabricWorld)) {
+            FabricWorld other = ((FabricWorld) o);
+            World otherWorld = other.worldRef.get();
+            World thisWorld = worldRef.get();
+            return otherWorld != null && otherWorld.equals(thisWorld);
+        } else if (o instanceof com.sk89q.worldedit.world.World) {
+            return ((com.sk89q.worldedit.world.World) o).getName().equals(getName());
```

Comment:
Wondering if we should never use a more specific check than this, i.e. always compare name. Currently this breaks the `equals` contract because it's not transitive.

Example:
```kotlin
val world1 = NativeWorld("foo")
val world2 = NativeWorld("foo")
val fbWorld1 = FabricWorld(world1)
val fbWorld2 = FabricWorld(world2)
val otherWorld = BukkitWorld(world1)
assertTrue(fbWorld1 == otherWorld)
assertTrue(otherWorld == fbWorld2)
assertTrue(fbWorld1 == fbWorld2) // fails!
```

Reply:
This one probably needs a separate discussion - most of these comments are about issues that have been copied from the Forge one, it may make sense to standardise or try using more common code.